File: apiutils/src/main/java/com/sky/http/DefaultSubScriber.java
package com.sky.http;


import com.google.gson.JsonSyntaxException;
import com.sky.SkyApp;
import com.sky.Common;
import com.sky.ErrorMes;
import com.sky.api.OnRequestCallback;
import com.sky.model.BaseEntity;

import io.reactivex.annotations.NonNull;
import io.reactivex.observers.DefaultObserver;
import retrofit2.HttpException;

/**
 * Created by SKY on 2017/5/29.
 */
public class DefaultSubScriber<T extends BaseEntity> extends DefaultObserver<T> {
    OnRequestCallback<T> request;

    public DefaultSubScriber(OnRequestCallback<T> request) {
        this.request = request;
    }

    @Override
    public void onNext(@NonNull T data) {
        if (request == null) return;
        if (data == null) {
            request.onFail(new ErrorMes(Common.NET_NULL, Common.NET_EMPTY));
        } else if (data.isSuccess()) {
            request.onSuccess(data);
        }else if (data.getStatus()==1) {
            request.onSuccess(data);
        } else if (Common.LOGIN == data.getCode()) {
            Common.getRxBus().send(Common.LOGIN);
        } else {
//            if (Common.DEBUG) SkyApp.getInstance().showToast(data.getMsg());
            request.onFail(new ErrorMes(data.getCode(), data.getMsg()));
        }
    }

    @Override
    public void onError(@NonNull Throwable e) {
        if (request == null) return;
        String error = "网络故障";
        if (e instanceof JsonSyntaxException) error = "服务器数据格式异常";
        else if (e instanceof HttpException) {
            error = e.getMessage();
        }

        request.onFail(new ErrorMes(Common.NET_NOT_FOUND,
                Common.DEBUG ? e.getMessage() : error));
        if (Common.DEBUG) SkyApp.getInstance().showToast(error);
    }

    @Override
    public void onComplete() {

    }
}

File: old/src/main/java/com/sky/chowder/utils/http/DefaultSubScriber.kt
package com.sky.chowder.utils.http

import com.google.gson.JsonSyntaxException
import com.sky.BuildConfig
import com.sky.oa.AppConstants
import com.sky.ErrorMes
import com.sky.SkyApp
import com.sky.chowder.api.OnRequestCallback
import com.sky.sdk.net.http.ApiResponse

import io.reactivex.annotations.NonNull
import io.reactivex.observers.DefaultObserver
import retrofit2.HttpException

/**
 * Created by SKY on 2017/5/29.
 */
class DefaultSubScriber<T : com.sky.sdk.net.http.ApiResponse<*>>(private var request: OnRequestCallback<T>?) : DefaultObserver<T>() {

    override fun onNext(@NonNull data: T) {
        request ?: return
        when {
            data == null -> request?.onFail(ErrorMes(com.sky.oa.AppConstants.NET_NULL, com.sky.oa.AppConstants.NET_EMPTY))
            data.isSuccess -> request?.onSuccess(data)
            data.status -> request?.onSuccess(data)
            else -> request?.onFail(ErrorMes(data.code, data.msg))
        }
    }

    override fun onError(@NonNull e: Throwable) {
        request ?: return
        var error = when {
            BuildConfig.DEBUG -> e.message
            e is JsonSyntaxException -> "服务器数据格式异常"
            e is HttpException -> e.message?:"服务器错误信息异常"
            else -> "网络故障"
        }
        request?.onFail(ErrorMes(com.sky.oa.AppConstants.NET_NOT_FOUND, error))
        if (BuildConfig.DEBUG) SkyApp.getInstance().showToast(error)
    }

    override fun onComplete() {

    }
}